exploration_recc_complaints/processing_recc_data.py

```diff
@@ -0,0 +1,207 @@
+
+    Args:
+        data: dataframe with RECC data
+    Returns:
+        The processed dataframe.
+    """
+
+    data["processed_complaint_summary"] = data["complaint_summary"].apply(deal_with_fit)
+
+    data["processed_complaint_summary"] = data[
+        "processed_complaint_summary"
+    ].str.lower()
+
+    data["processed_complaint_summary"] = data["processed_complaint_summary"].apply(
+        merges_expression_variations
+    )
+
+    data["complaint_length"] = data["processed_complaint_summary"].str.len()
+
+    # there are other ways we could do this
+    data["tokens"] = data["processed_complaint_summary"].apply(
+        lambda x: re.sub("[^A-Za-z0-9]+", " ", x)
+    )
+
+    data["tokens"] = data["tokens"].apply(lambda x: word_tokenize(x))
+
+    data["stems"] = data["tokens"].apply(stemming)
+
+    data["lemmas"] = data["tokens"].apply(lemmatising)
+
+    return data
+
+
+def create_dummy_variables_and_total(
+    data: pd.DataFrame, variable: str, prefix: str
+) -> pd.DataFrame:
+    """
+    Creates dummy variables based on an existing variable in the data frame.
+    Also creates a variable with the totals (number of dummy variables with 1 for each instance).
+
+    Args:
+        data: dataframe with RECC data
+        variable: variable to split into dummy variables
+        prefix: prefix to add to each dummy variable
+    Returns:
+        The processed dataframe.
+
+    """
+
+    unique_values = list(set(data[variable].unique().sum().split("; ")[:-1]))
+
+    dummy_variable_names = [prefix + ":" + c for c in unique_values]
+
+    for c in range(len(unique_values)):
+        data[dummy_variable_names[c]] = data[variable].apply(
+            lambda x: 1 if unique_values[c] in x else 0
+        )
+
+    data["number_of_" + variable] = data[dummy_variable_names].sum(axis=1)
+
+    return data
+
+
+def extract_info_from_date(data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Extracts info such as year and month from the date variable.
+
+    Args:
+        data: dataframe with RECC data
+    Returns:
+        The processed dataframe.
+    """
+    data["year_month"] = data["date_received"].apply(lambda x: x[:7])
+    data["date_received"] = data["date_received"].apply(
+        lambda x: datetime.strptime(x, "%Y-%m-%d")
+    )
+    data["year"] = data["date_received"].dt.year
+    data["month"] = data["date_received"].dt.month
+
+    return data
+
+
+def changes_to_categories(data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Processes the categories column by filling NAs with "No category specified; "
+    and adding "; " at the end of each categories' instance to make it easier to split by category in the future.
+    Also creates an additional variable with a short version of each category.
+
+    Args:
+        data: dataframe with RECC data
+    Returns:
+        Processed dataframe.
+    """
+    data["categories"].fillna("No category specified; ", inplace=True)
```

Comment:
Not for tech columns, as this function is only applied to categories. But it will have `number_of_categories` as 1 if "No category specified; ", so I'll fix that so that it is 0. Thanks for spotting it @ch-williamson  👍

Reply:
Also added plots to look at these distributions under the descriptive analysis script:

- percentage of complaints with a given number of categories (`percentage_complaints_by_number_of_categories.png`)
- percentage of complaints with a given number of technologies (`percentage_complaints_by_number_of_technologies.png`)